apps/docs/pages/guides/database/postgres/row-level-security.mdx

```diff
@@ -0,0 +1,378 @@
+
+Policies are "implicit where clauses", so it's common to run `select` statements without any filters. This is a bad pattern for performance. Instead of doing this (JS client example):
+
+{/* prettier-ignore */}
+```js
+const { data } = supabase
+  .from('table')
+  .select()
+```
+
+You should always add a filter:
+
+{/* prettier-ignore */}
+```js
+const { data } = supabase
+  .from('table')
+  .select()
+  .eq('user_id', userId)
+```
+
+Even though this duplicates the contents of the Policy, Postgres can use the to construct a better query plan.
+
+#### Benchmarks
+
+| Test                                                                                              | Before (ms) | After (ms) | % Improvement | Change                                                                                                                                 |
+| ------------------------------------------------------------------------------------------------- | ----------- | ---------- | ------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
+| [test3-addfilter](https://github.com/GaryAustin1/RLS-Performance/tree/main/tests/test3-addfilter) | 171         | 9          | 94.74%        | <details className="cursor-pointer">Before:<br/>`auth.uid() = user_id`<br/><br/>After:<br/>add `.eq` or `where` on `user_id`</details> |
+
+### Use security definer functions
+
+A "security definer" function runs using the same role that _created_ the function. This means that if you create a role with a superuser (like `postgres`), then that function will have `bypassrls` privileges. For example, if you had a policy like this:
+
+```sql
+create policy "rls_test_select" on test_table
+to authenticated
+using (
+  exists (
+    select 1 from roles_table
+    where auth.uid() = user_id and role = 'good_role'
+  )
+);
+```
+
+We can instead create a `security definer` function which can scan `roles_table` without any RLS penalties:
+
+```sql
+create function private.has_good_role()
+returns boolean
+language plpgsql
+security definer -- will run as the creator
+as $$
+begin
+  return exists (
+    select 1 from roles_table
+    where auth.uid() = user_id and role = 'good_role'
+  )
+end;
+$$;
+
+-- Update our policy to use this function:
+create policy "rls_test_select"
+on test_table
+to authenticated
+using ( private.has_good_role() );
+```
+
+{/* prettier-ignore */}
+<Admonition type="caution">
+Security-definer functions should never be created in a schema in the "Exposed schemas" inside your [API settings](https://supabase.com/dashboard/project/_/settings/api)`. 
+</Admonition>
+
+### Minimize joins
+
+You can often rewrite your Policies to avoid joins between the source and the target table. Instead, try to organize your Policy to fetch all the relevant data from the target table into an array or set, then you can use an `IN` or `ANY` operation in your filter.
```

Comment:
```suggestion
You can often rewrite your Policies to avoid joins between the source and the target table. Instead, try to organize your policy to fetch all the relevant data from the target table into an array or set, then you can use an `IN` or `ANY` operation in your filter.
```